[DPE-9311] Initial user documentation (#24)
canonical/valkey-operator: docs/tutorial.md

(tutorial)=

> # Tutorial

Reviewer: Praise: I like the single-page approach and the simple page title.

Author: Thank you!

This hands-on tutorial aims to help you learn how to deploy Charmed Valkey and 
become familiar with its available operations.

> ## Prerequisites

Reviewer: Note: I was expecting to see some software and/or hardware requirements here. If there are any, consider adding them here. For example: CPU, RAM, Disk, Ubuntu version, etc. Non-blocking for sure.

Author: No prerequisites for this charm. For production setup, we will have a separate page at some point.

Reviewer: The prerequisites are for the tutorial. For example, you have `multipass launch --cpus 4 --memory 8G --disk 50G` which gives you a great idea of what to put here.

Reviewer: I see that no changes were introduced to add the hardware and software (at least Ubuntu and its version) requirements. Can you please comment on this idea?

Author: I've misunderstood your comment in the first place. As for the requirements for the tutorial, I've added a section below. Note that I've also reduced the required resources to lower the bar for users who want to try it on their own machine.

While this tutorial intends to guide you as you deploy Charmed Valkey for the
first time, it will be most beneficial if:
- You have some experience using a Linux-based CLI
- You are familiar with the [Juju orchestration engine](https://documentation.ubuntu.com/juju/latest/)

> ---

Reviewer: Todo: We no longer use this syntax. It was for a horizontal line, separating a Summary/ Table of contents from the main content. We have a local ToC now on the right, so we don't put it inside the page's content, so we don't use the horizontal lines anymore for anything.

```suggestion
```

Author: removed

Check the IP addresses associated with each application unit with the `juju status` command: 

```text
Unit       Workload  Agent  Address      Ports  Message
valkey/0*  active    idle   10.1.44.126         
```

The user we will connect to in this tutorial will be the internal `charmed-operator`
user of Charmed Valkey. To retrieve its associated password, run the following command:

```text
juju show-secret valkey-peers.valkey.app.internal_users_secret --reveal
```

Copy the content displayed for `charmed-operator-password`.

### Access the database

The easiest way to interact with Valkey is via [its command line interface `valkey-cli`](https://valkey.io/topics/cli/).
which can be installed with the [`valkey-tools` package in Ubuntu](https://packages.ubuntu.com/search?suite=noble&section=all&arch=any&keywords=valkey-tools&searchon=names):

```text
sudo apt update && sudo apt install valkey-tools
```

Run the command below to connect to your Charmed Valkey database, using the host's IP address:

```text
valkey-cli -h 10.1.44.126 -p 6379
```

Run the following command to log in and perform a basic health check, using the
previously retrieved credentials:

```text

> 10.1.44.126:6379> AUTH charmed-operator <your-password-here>

Reviewer: Chore: Please make sure to fix code blocks (input/output combinations and cli prompts).

Author: done

OK
10.1.44.126:6379> ping
PONG
```

Now it is possible to perform Valkey commands on the database. To set a key `mykey`

> to the value `HelloWorld`, run this command:

Reviewer: Suggestion: Wording improvement - deleting redundancy.

```suggestion
to the value `HelloWorld`:
```

Author: updated

(tutorial)=
# Tutorial

This hands-on tutorial aims to help you learn how to deploy Charmed Valkey and 
become familiar with its available operations.

## Prerequisites

While this tutorial intends to guide you as you deploy Charmed Valkey for the
first time, it will be most beneficial if:
- You have some experience using a Linux-based CLI
- You are familiar with the [Juju orchestration engine](https://documentation.ubuntu.com/juju/latest/)

## Set up the environment

This tutorial will deploy Charmed Valkey on Kubernetes.

First, we will set up a cloud environment using [Multipass](https://multipass.run/)
with [MicroK8s](https://microk8s.io/docs) and [Juju](https://documentation.ubuntu.com/juju/3.6/). 
This is the quickest and easiest way to get your machine ready for using Charmed Valkey
on Kubernetes. 

> 

Reviewer 2: Add a note here that Valkey is dual substrate etc (ref previous comment)

Author: Updated.

(tutorial)=
# Tutorial

This hands-on tutorial aims to help you learn how to deploy Charmed Valkey and 
become familiar with its available operations.

## Prerequisites

While this tutorial intends to guide you as you deploy Charmed Valkey for the
first time, it will be most beneficial if:

> - You have some experience using a Linux-based CLI

Reviewer: Suggestion: Please make sure to have empty lines before and after all lists.

```suggestion

- You have some experience using a Linux-based CLI
```

- You are familiar with the [Juju orchestration engine](https://documentation.ubuntu.com/juju/latest/)

## Set up the environment

Charmed Valkey can be deployed on both Kubernetes and VM clouds. For this tutorial,
we will deploy Charmed Valkey on Kubernetes using MicroK8s.

First, we will set up a cloud environment using [Multipass](https://multipass.run/)
with [MicroK8s](https://microk8s.io/docs) and [Juju](https://documentation.ubuntu.com/juju/3.6/). 
This is the quickest and easiest way to get your machine ready for using Charmed Valkey
on Kubernetes. If you are interested in VM deployments, [LXD](https://documentation.ubuntu.com/lxd/latest/)
can be used instead of MicroK8s. 

### Multipass

[Multipass](https://multipass.run/) is a quick and easy way to launch virtual
machines running Ubuntu. It uses the [cloud-init](https://cloud-init.io/) standard
to install and configure all the necessary parts automatically.

Install Multipass from the [snap store](https://snapcraft.io/multipass):

> ```shell

Reviewer: ```suggestion

```shell
```

sudo snap install multipass
```

Spin up a new VM using [`multipass launch`](https://multipass.run/docs/launch-command)
with the [charm-dev](https://github.com/canonical/multipass-blueprints/blob/main/v1/charm-dev.yaml)
cloud-init configuration:

```shell
multipass launch --cpus 4 --memory 8G --disk 50G --name dev-vm charm-dev
```

As soon as a new VM has started, access it:

```shell
multipass shell dev-vm
```

```{tip}
If at any point you'd like to leave a Multipass VM, enter `Ctrl+D` or type `exit`.
```

All necessary components have been pre-installed inside the VM already, like LXD and Juju.

### Juju

Let's bootstrap Juju to use the local MicroK8s controller. We will call it 
"k8s-controller", but you can give it any name you'd like:

```shell
juju bootstrap microk8s k8s-controller
```

If you decided to deploy an LXD environment instead of MicroK8s, bootstrap the LXD
controller with this command:

```shell

> juju bootstrap localhost lxd-controller

Reviewer: Suggestion: This command is an alternative to the previous one. We should be extra careful to prevent users running both. Maybe we can put the LXD instructions or just the command in an expandable block? See [syntax example](https://github.com/canonical/kafka-operator/blob/df663a0510670f60d4351650fe5ead742f3dcbb9/docs/tutorial/environment.md?plain=1#L74-L86).

Author: Nice idea, I like that!

You can track the progress by running:

```shell
watch juju status --color
```

This command is useful for checking the real-time information about the state
of a charm and the machines hosting it. Check the [`juju status` documentation](https://juju.is/docs/juju/juju-status)
for more information about its usage.

When the application is ready, `juju status` will show something similar to the
sample output below:

```text
Model     Controller      Cloud/Region        Version  SLA          Timestamp
tutorial  k8s-controller  microk8s/localhost  3.6.14   unsupported  12:53:52+01:00

App     Version  Status  Scale  Charm   Channel  Rev  Address         Exposed  Message
valkey           active      1  valkey  9/edge    11  10.152.183.123  no       

Unit       Workload  Agent  Address      Ports  Message
valkey/0*  active    idle   10.1.44.126         
```

You can also watch juju logs with the [`juju debug-log`](https://juju.is/docs/juju/juju-debug-log) command.

## Access Valkey

In this section, you will learn how to get the credentials of your deployment and
connect to the Valkey database.

```{caution}
This part of the tutorial accesses Valkey via the `charmed-operator` user. 

> **Do not directly interface with the `charmed-operator` user in a production environment.**

Reviewer: Suggestion: Wording improvements.

```suggestion
This part of the tutorial uses the charmed-operator user to access Valkey.

**Do not use the charmed-operator user directly in a production environment.**
```

Reviewer: Question: Can we add some guidance on what to do instead for a production environment? It's OK if we can't because we don't have such content or feature yet.

```

### Retrieve credentials

Connecting to the database requires that you know two pieces of information: 

> 1. The internal Valkey database user credentials (username and password)

Reviewer: ```suggestion

1. The internal Valkey database user credentials (username and password)
```

2. The host machine's IP address. 

Check the IP addresses associated with each application unit with the `juju status` command: 

```text
Unit       Workload  Agent  Address      Ports  Message
valkey/0*  active    idle   10.1.44.126         
```

The user we will connect to in this tutorial will be the internal `charmed-operator`
user of Charmed Valkey. To retrieve its associated password, run the following command:

```shell
juju show-secret valkey-peers.valkey.app.internal_users_secret --reveal
```

Copy the content displayed for `charmed-operator-password`.

### Access the database

The easiest way to interact with Valkey is via [its command line interface `valkey-cli`](https://valkey.io/topics/cli/).
which can be installed with the [`valkey-tools` package in Ubuntu](https://packages.ubuntu.com/search?suite=noble&section=all&arch=any&keywords=valkey-tools&searchon=names):

```shell
sudo apt update && sudo apt install valkey-tools
```

> Run the command below to connect to your Charmed Valkey database, using the host's IP address:

Reviewer: Suggestion: let's be a bit more specific here.

```suggestion
Run the command below to connect to your Charmed Valkey database, using the unit's IP address:
```

```shell
valkey-cli -h 10.1.44.126 -p 6379
```

In Kubernetes environments, it is also possible to connect to the database using the DNS name
of the unit. For that, connect into the unit:

```shell
juju ssh --container valkey valkey/0
```

Use the DNS endpoint to access Valkey from within the unit:

```shell
valkey-cli -h valkey-0.valkey-endpoints -p 6379
```

Run the following command to log in, using the previously retrieved credentials:

```text

> 10.1.44.126:6379> AUTH charmed-operator <your-password-here>

Reviewer: Suggestion: Let's fix the CLI prompts here and after, as mentioned earlier.

```

Now perform a basic health check with this command:

```shell
10.1.44.126:6379> ping
```

You should receive this response from the Valkey server:

```text
PONG
```

Now it is possible to perform Valkey commands on the database. To set a key `mykey`
to the value `HelloWorld`:

```shell
10.1.44.126:6379> set mykey "HelloWorld"
```

In order to retrieve the key you just set, run the following command:

```shell

> 10.1.44.126:6379> get mykey

Reviewer: praise: I like the simple test we have here.

client TLS. Valkey now listens on port `6380` (previously it was `6379`).

Use `openssl` to connect to Valkey and check the TLS certificate in use. Note 
that your unit's IP address will likely be different to the one shown below:

```shell
openssl s_client -connect 10.1.44.126:6380  2>/dev/null | openssl x509 -noout -dates
```

You should see the validity dates of the certificate in use as output:

```text
notBefore=Mar 16 13:52:57 2026 GMT
notAfter=Jun 14 13:52:57 2026 GMT
```

Congratulations! Valkey is now using a TLS certificate generated by the external
application `self-signed-certificates`.

To remove the external TLS, remove the integration:

```shell
juju remove-relation valkey:client-certificates self-signed-certificates:certificates
```

The Charmed Valkey application is not using TLS anymore.

## Clean up your environment

In this tutorial we've successfully deployed Valkey on MicroK8s, added and removed
cluster members, and enabled a layer of security with TLS.

You may now keep your Charmed Valkey deployment running and write to the database 
or remove it entirely using the steps in this page. 

> If you'd like to keep your environment for later, simply stop your VM with

Reviewer: Suggestion: I'd add a small sentence about leaving the VM's shell with Ctrl+D, just to remind users about it. But it might be a bit too dumbing down. So, up to you.

machines running Ubuntu. It uses the [cloud-init](https://cloud-init.io/) standard
to install and configure all the necessary parts automatically.

Install Multipass from the [snap store](https://snapcraft.io/multipass):

```shell
sudo snap install multipass
```

Spin up a new VM using [`multipass launch`](https://multipass.run/docs/launch-command)
with the [charm-dev](https://github.com/canonical/multipass-blueprints/blob/main/v1/charm-dev.yaml)
cloud-init configuration:

```shell
multipass launch --cpus 4 --memory 8G --disk 50G --name dev-vm charm-dev
```

As soon as a new VM has started, access it:

```shell
multipass shell dev-vm
```

```{tip}
If at any point you'd like to leave a Multipass VM, enter `Ctrl+D` or type `exit`.
```

All necessary components have been pre-installed inside the VM already, like LXD and Juju.

### Juju

Let's bootstrap Juju to use the local MicroK8s controller. We will call it 
"k8s-controller", but you can give it any name you'd like:

```shell
juju bootstrap microk8s k8s-controller
```

If you decided to deploy an LXD environment instead of MicroK8s, bootstrap an LXD
controller instead.

> <details> <summary> Bootstrap example</summary>

Reviewer: Suggestion: Let's make this visible link text LXD specific, since we already have a bootstrap command above just for K8s.

Author: updated
